src/main/java/seedu/address/model/VersionedAddressBook.java

```diff
@@ -0,0 +1,156 @@
+package seedu.address.model;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import javafx.collections.ObservableList;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.person.Person;
+
+/**
+ * Represents the different states of the Address Book for the user.
+ */
+public class VersionedAddressBook extends AddressBook {
+    public static final String MESSAGE_NO_MORE_HISTORY = "This is the earliest version that user can retrieve";
+    public static final String MESSAGE_UNSAVED_CHANGES = "There are unsaved changes in the current state. "
+            + "Please commit or discard the changes before undoing.";
+    private final ArrayList<AddressBook> addressBookStateList;
+    private int currentStatePointer;
+    private AddressBook current;
+
+
+    /**
+     * Constructs an empty state list for the address book.
+     */
+    public VersionedAddressBook() {
+        addressBookStateList = new ArrayList<>();
+        currentStatePointer = -1;
+        current = new AddressBook();
+        this.commitAddressBook();
+    }
+
+    /**
+     * Constructs a state list for the address book and saves the initial copy.
+     */
+    public VersionedAddressBook(ReadOnlyAddressBook addressBook) {
+        addressBookStateList = new ArrayList<>();
+        currentStatePointer = -1;
+        current = new AddressBook(addressBook);
+        this.commitAddressBook();
+    }
+
+    /**
+     * Saves the whole address book state in the history.
+     */
+    public void commitAddressBook() {
+        currentStatePointer++;
+        addressBookStateList.subList(currentStatePointer, addressBookStateList.size()).clear();
+        addressBookStateList.add(currentStatePointer, new AddressBook(current));
+    }
+
+    /**
+     * Reverts the given AddressBook to the previous state.
+     */
+    public void undoAddressBook() throws CommandException {
+        if (currentStatePointer <= 0) {
+            throw new CommandException(MESSAGE_NO_MORE_HISTORY);
+        }
+
+        if (!current.equals(addressBookStateList.get(currentStatePointer))) {
+            throw new CommandException(MESSAGE_UNSAVED_CHANGES);
+        }
```

Comment:
Actually not only `resetData`, all uncommitted call to other data-changing methods like `addPerson()` , `removePerson()` etc. will trigger this. 
However, I think using assert would be a better choice here because an unsaved undo is our programmer's fault (probably because we forget to call commit after command execution), but I changed to exception for easier testing. Maybe there is a better way!